src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -169,6 +197,47 @@ class thread_priority {
   long task_priority() const { return std::numeric_limits<long>::max() - (task_id + 1); }
 };
 
+/**
+ * Holds metrics for a given task/thread about retry counts and times. It is here
+ * because the mapping between tasks and threads can be complicated and can span
+ * different time ranges too.
+ */
+class task_metrics {
+  public:
+  // metric for being able to report how many times each type of exception was thrown,
+  // and some timings
+  int num_times_retry_throw       = 0;
+  int num_times_split_retry_throw = 0;
+  long time_blocked_nanos         = 0;
+  // The amount of time that this thread has lost due to retries (not including blocked time)
+  long time_lost_nanos = 0;
+
+  void take_from(task_metrics& other) {
+    this->num_times_retry_throw += other.num_times_retry_throw;
+    other.num_times_retry_throw = 0;
+    this->num_times_split_retry_throw += other.num_times_split_retry_throw;
+    other.num_times_split_retry_throw = 0;
+    this->time_blocked_nanos += other.time_blocked_nanos;
+    other.time_blocked_nanos = 0;
+    this->time_lost_nanos += other.time_lost_nanos;
+    other.time_lost_nanos = 0;
```

Comment:
That is very cumbersome. Let use this:
```suggestion
  void take_from(task_metrics& other) {
    *this = other;
    other = task_metrics{};
```

Reply:
Or we can just ignore `other` to let it keep its original values.